### PR TITLE
Add SCAN_METHOD feature and image caption replacement

### PR DESCRIPTION
**Pull request by mguyard**
Add SCAN_METHOD feature (to allow a checkbox to define which page nee…d to be analyze) and usage of image caption to replace ocr_text in caption and not only add new paragraph in end of page

As purpose in https://github.com/MichielvanBeers/notion-auto-ocr/issues/1

### CODE REVIEW

**Comment by MichielvanBeers** [comment]
@mguyard first of all: wow! Great work on not only adding the functionality but also documenting it very well. Really awesome to see this 🤩!

I've pulled the PR and everything works great, including the error handling. I did find a limitation with the caption though: when there are more than 2000 characters in the text, the Notion API give the following error: ```{"object":"error","status":400,"code":"validation_error","message":"body failed validation: body.image.caption[0].text.content.length should be ≤ `2000`, instead was `2375`."}``` This seems to be [a limitation of the number of characters per block in Notion](https://guide.unito.io/a-primer-on-unitos-notion-integration). I used [this image](https://miro.medium.com/max/720/0*dUImW58g7_K-5j3P) for testing.

I was thinking that we might want to write some information in the caption if this is the case, for example, "This image contains more than 2000 characters, please put the ocr_text as a tag below the picture instead of in the caption."

What do you think?

NB: I'll put some minor comments on styling conventions, just to polish things up

File: app.py [hunk 1]
```diff
@@ -13,47 +13,80 @@
...
 
+# This is a dictionary that contains the headers that will be sent with every request to the Notion
+# API.
 HEADERS = {
     "Authorization": "Bearer " + NOTION_TOKEN,
     "Content-Type": "application/json",
     "Notion-Version": "2022-06-28"
 }
 
+def get_scan_requestBody():
```
Comment (MichielvanBeers) [inline app.py]: Suggesting to make this more 'pythonic' by changing the camelCase into a snake_case (eg get_scan_request_body)

File: app.py [hunk 2]
```diff
@@ -13,47 +13,80 @@
...
+                    },
+                    "sorts": [
+                        {
+                            "property": "Created time",
+                            "direction": "descending"
+                        }
+                    ]
+                }
+        case _:
+            print(f"Scan method parameter is invalid. Valid values are checkbox or createtime. You have define", SCAN_METHOD)
```
Comment (MichielvanBeers) [inline app.py]: Small typo: "You have define" --> "You have defined:"

File: app.py [hunk 3]
```diff
@@ -13,47 +13,80 @@
...
-                }
-            },
-            "sorts": [
-                {
-                    "property": "Created time",
-                    "direction": "descending"
-                }
-            ]
-        }
+    request_body = get_scan_requestBody()
```
Comment (MichielvanBeers) [inline app.py]: Suggesting to make this more 'pythonic' by changing the camelCase into a snake_case (eg get_scan_request_body)

File: app.py [hunk 4]
```diff
@@ -89,22 +138,36 @@ def get_images_to_scan_in_page(page_id, headers):
 
     for index, block in enumerate(results):
         if block['type'] == 'image':
+            if block['image']['caption']:
+                for indexCaption, caption in enumerate(block['image']['caption']):
```
Comment (MichielvanBeers) [inline app.py]: Suggesting to make this more 'pythonic' by changing the camelCase into a snake_case (eg index_caption)

File: app.py [hunk 5]
```diff
@@ -89,22 +138,36 @@ def get_images_to_scan_in_page(page_id, headers):
...
+                    if caption['plain_text'] and 'ocr_text' in caption['plain_text'].split('\n'):
+                        caption_index = indexCaption
```
Comment (MichielvanBeers) [inline app.py]: Suggesting to make this more 'pythonic' by changing the camelCase into a snake_case (eg index_caption)

File: app.py (hunk 1 above)
Comment (mguyard) [reply app.py]: Yes I don't develop in python anymore but in nodejs and so I use camelCase instead.
So i've working hard to use snake_case but sometime my daemons were to strong :)

**Comment by mguyard** [comment]
> I did find a limitation with the caption though: when there are more than 2000 characters in the text, the Notion API give the following error: `` {"object":"error","status":400,"code":"validation_error","message":"body failed validation: body.image.caption[0].text.content.length should be ≤ `2000`, instead was `2375`."} `` This seems to be [a limitation of the number of characters per block in Notion](https://guide.unito.io/a-primer-on-unitos-notion-integration). I used [this image](https://miro.medium.com/max/720/0*dUImW58g7_K-5j3P) for testing.
> 
> I was thinking that we might want to write some information in the caption if this is the case, for example, "This image contains more than 2000 characters, please put the ocr_text as a tag below the picture instead of in the caption."

I will try to find a solution. We are able to have multiple captions. Sl instead of putting everything in the same caption, i try to split text on the last space of each part of 2000 charaters and put in different caption.
As explained above, python is far behind me so if you have some suggestion to do this you are welcome. I think i will use rindex/rfind to find last space by 2000 part

**Comment by MichielvanBeers** [comment]
That is even better! I'm trying something with regex to clean up the last broken word after 2000 characters, but not super happy with the result yet. I'll let you know if I have anything working!

**Comment by MichielvanBeers** [comment]
Maybe this could work. What do you think?

```
import re

# Would be set to 2000 in implementation
MAX_CHARACTERS = 20
string = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin non pretium augue. Sed scelerisque diam dolor, nec pellentesque sapien feugiat nec. Sed tortor nulla, auctor ut fringilla eu, commodo id elit. "
array_string = []

# Check if string longer than MAX_CHARACTERS, otherwise just continue
while len(string) >= MAX_CHARACTERS:

    # Cut the string to MAX_CHARACTERS
    sub_string = string[:MAX_CHARACTERS] 

    # Match the last characters + the first space
    cleaned_sub_string = re.sub("\s\S*$","",sub_string)
    
    # Add that to an array
    array_string.append(cleaned_sub_string)

    # Remove the string that has been added to the array, from the orginal string
    string = string.replace(cleaned_sub_string,"")

else:
    array_string.append(string)
```

**Comment by mguyard** [comment]
I've push a new version who manage the 2000 charaters limitations.
After some test, all seems to works. But if you find a bug inform me.

I've also corrected a issue (seems to only apply on my addon) when on the same page you have multiple image but one of push don't have success, checkbox was move to False. So now, i change checkbox only if all image are in success (as checkbox is on the page and not on image).

**Comment by MichielvanBeers** [comment]
Great! I'll make sure to review and test everything tomorrow. I'll keep you posted

**Comment by MichielvanBeers** [comment]
@mguyard reviewed and tested! Everything works as expected :). Also tested if the combination of caption and paragraph tags works and it does!

Once again, thanks for all the work and documentation! If all is well, a new docker image should be available in a couple of minutes (I'll make sure to test that in my environment as well

**Pull request merged**